pkg/gameserverallocations/allocator.go

```diff
@@ -128,6 +132,19 @@ func NewAllocator(policyInformer multiclusterinformerv1alpha1.GameServerAllocati
 	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
 	ah.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "GameServerAllocation-Allocator"})
 
+	// Find Allocator IPs with retry
```

Comment:
This is to find the allocator service IP addresses at the start up, so that for multi-cluster allocation, if the allocation endpoint is the same as the current cluster endpoint, treat it as local allocation vs allocating from remote cluster.

Reply:
Don't we know this from our CRD configuration? Anything that's defined as an AllocationEndpoint in the policy would be considered external?

Reply:
> Don't we know this from our CRD configuration? 

Which CRD configuration? The service CRD has this information at service.Status.LoadBalancer.Ingress[0].IP. That's the one we use to find the current cluster's agones-allocator endpoint. 

> Anything that's defined as an AllocationEndpoint in the policy would be considered external?

The multi-cluster allocation policy may have a policy to allocate from the current cluster with the highest priority and if it cannot allocate, then allocate from the cluster with the second highest priority. To avoid additional allocation request hop when the cluster should self-allocate base on the allocation request and policy, I have added a condition to self-allocate if the allocation policy's IP and namespace belongs to the same cluster and namespace that has received the allocation request.

Reply:
The GameServerAllocationPolicy CRD - is the ClusterConnectionInfo only external endpoints? Or does that include itself as well? If so, wouldn't it be much easier to be able to delineate that as local within the cluster?

If it's self allocating, isn't that a request through the `GameServerAllocation` ? What endpoint does it need?

There are also multiple things here in this PR, I'm struggling to put together a coherent picture of what all this does, unfortunately :disappointed: maybe I'm getting confused.

I just feel like we're making assumptions about what IPs are where, when maybe we shouldn't - it could paint us into a corner we can't get out of.

Reply:
> The GameServerAllocationPolicy CRD - is the ClusterConnectionInfo only external endpoints? Or does that include itself as well? If so, wouldn't it be much easier to be able to delineate that as local within the cluster?
> 
It depends. We can make the assumption that if endpoint is set to 127.0.0.1 or localhost, then it is self-allocating and ignore if a user set the AllocationEndpoint to the allocator IP that belongs to the same cluster, but I am not a fan of making assumptions and expecting users do the right thing. WDYT?

> If it's self allocating, isn't that a request through the `GameServerAllocation` ? What endpoint does it need?
> 
There is a MultiClusterSetting that enables or disables the multi-cluster feature through allocation request. MatchMaker is not supposed to know about the allocation policy to enable or disable the feature.

> There are also multiple things here in this PR, I'm struggling to put together a coherent picture of what all this does, unfortunately  maybe I'm getting confused.
> 
Sorry about that. It was a small focused PR that did not get traction and then I had more changes rebased on that; so I included the new changes as well. I can break it to two PR if it helps.

> I just feel like we're making assumptions about what IPs are where, when maybe we shouldn't - it could paint us into a corner we can't get out of.

What do you recommend?

Reply:
I gave it more thoughts. You have a point. I made a change that if the AllocationEndpoints is not specified, then allocate locally.